cpp/translator.cc

```diff
@@ -322,11 +351,45 @@ void GetVuFindTranslationsAsHTMLRowsFromDatabase(DbConnection &db_connection, co
 }
 
 
+void SetupKeyWordSortLimitQuery(DbConnection &db_connection, std::string * const create_sort_limit, const std::string &offset,
+                                const bool use_untranslated_filter)
+{
+    // The LIMIT parameter can only work with constants, but we want entries per page to be lines, i.e. german translations in our table
+    // so we have to generate a dynamic limit using temporary tables
+    if (use_untranslated_filter) {
+        const std::string translator=GetTranslatorOrEmptyString();
+        if (translator.empty()) {
+            ShowErrorPage("Error - No Valid User", "Not valid user selected");
+            std::exit(0);
+        }
+
+        const std::string translated_by_translator("CREATE TEMPORARY TABLE translated_by_translator (INDEX (ppn)) AS "
+                                                   "(SELECT DISTINCT ppn FROM keywords_ger_sorted WHERE translator=\'" + translator + "\')");
+        db_connection.queryOrDie(translated_by_translator);
+
+        // Get all PPNs that are so far untouched by this translator
+        const std::string untranslated_by_translator_ger_sorted("CREATE TEMPORARY TABLE untranslated_by_translator_ger_sorted AS "
+                                                     "(SELECT DISTINCT l.ppn FROM (SELECT ppn FROM keywords_ger_sorted "
+                                                     "WHERE language_code='ger' ORDER BY translation) AS l "
+                                                     "LEFT JOIN translated_by_translator AS r "
+                                                     "ON l.ppn=r.ppn WHERE r.ppn IS NULL)");
+        db_connection.queryOrDie(untranslated_by_translator_ger_sorted);
+
+        *create_sort_limit = "CREATE TEMPORARY TABLE sort_limit AS (SELECT DISTINCT ppn FROM  untranslated_by_translator_ger_sorted "
+                             "LIMIT " +  offset + ", " + std::to_string(ENTRIES_PER_PAGE) + ")";
+    } else {
+        *create_sort_limit = "CREATE TEMPORARY TABLE sort_limit AS (SELECT ppn FROM keywords_ger_sorted "
+                                        "WHERE language_code='ger' ORDER BY translation  LIMIT " + offset + ", "
+                                        + std::to_string(ENTRIES_PER_PAGE) +  ")";
+    }
+}
+
+
 void GetKeyWordTranslationsAsHTMLRowsFromDatabase(DbConnection &db_connection, const std::string &lookfor,
                                                   const std::string &offset, std::vector<std::string> *const rows,
                                                   std::string *const headline,
                                                   const std::vector<std::string> &translator_languages,
-                                                  const std::vector<std::string> &additional_view_languages) {
+                                                  const std::vector<std::string> &additional_view_languages, bool use_untranslated_filter) {
     rows->clear();
```

Comment:
Unless you modify "use_untranslated_filter" in this function, please make this parameter const.